File: sdk/typescript/src/namespaces/evolution.ts
/**
 * Evolution Namespace API
 *
 * Provides access to agent prompt evolution, patterns, A/B testing, and summaries.
 */

export interface EvolutionPattern {
  id: string;
  pattern_type: string;
  description: string;
  frequency: number;
  impact_score?: number;
  agents?: string[];
  metadata?: Record<string, unknown>;
}

export interface EvolutionSummary {
  total_agents: number;
  total_evolutions: number;
  top_patterns: EvolutionPattern[];
  improvement_rate?: number;
  metadata?: Record<string, unknown>;
}

export interface EvolutionHistoryEntry {
  timestamp: string;
  version?: number;
  score?: number;
  changes?: Record<string, unknown>;
  trigger?: string;
}

export interface EvolutionHistoryResponse {
  agent: string;
  history: EvolutionHistoryEntry[];
  total: number;
}

export interface PromptVersion {
  agent: string;
  version: number;
  prompt: string;
  created_at: string;
  performance_score?: number;
  metadata?: Record<string, unknown>;
}

export interface ABTest {
  id: string;
  agent: string;
  variant_a: string;
  variant_b: string;
  status: 'active' | 'completed' | 'cancelled';
  started_at: string;
  completed_at?: string;
  results?: {
    variant_a_score: number;
    variant_b_score: number;
    winner?: 'a' | 'b' | 'tie';
    sample_size: number;
    confidence: number;
  };
}

export interface CreateABTestRequest {
  agent: string;
  variant_a: string;
  variant_b: string;
  config?: Record<string, unknown>;
}

export interface AgentEvolutionTimelineOptions {
  limit?: number;
  offset?: number;
}

export interface AgentEvolutionEloTrendsOptions {
  period?: '24h' | '7d' | '30d' | '90d' | string;
}

interface EvolutionClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; body?: unknown }
  ): Promise<T>;
}

export class EvolutionAPI {
  constructor(private client: EvolutionClientInterface) {}

  /**
   * Get top evolution patterns across all agents.
   */
  async getPatterns(params?: {
    limit?: number;
  }): Promise<{ patterns: EvolutionPattern[] }> {
    return this.client.request('GET', '/api/v1/evolution/patterns', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Get evolution summary statistics.
   */
  async getSummary(): Promise<EvolutionSummary> {
    return this.client.request('GET', '/api/v1/evolution/summary');
  }

  /**
   * Get prompt evolution history for an agent.
   */
  async getHistory(agent: string): Promise<EvolutionHistoryResponse> {
    return this.client.request('GET', `/api/v1/evolution/${agent}/history`);
  }

  /**
   * Get the current or a specific prompt version for an agent.
   */
  async getPrompt(agent: string, params?: {
    version?: number;
  }): Promise<PromptVersion> {
    return this.client.request('GET', `/api/v1/evolution/${agent}/prompt`, {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * List all A/B tests.
   */
  async listABTests(params?: {
    limit?: number;
    status?: string;
  }): Promise<{ tests: ABTest[]; total?: number }> {
    return this.client.request('POST', '/api/v1/evolution/ab-tests', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Get the active A/B test for an agent.
   */
  async getActiveABTest(agent: string): Promise<ABTest | null> {
    return this.client.request('GET', `/api/v1/evolution/ab-tests/${agent}/active`);
  }

  /**
   * Start a new A/B test.
   */
  async createABTest(body: CreateABTestRequest): Promise<ABTest> {
    return this.client.request('POST', '/api/v1/evolution/ab-tests', { body });
  }

  /**
   * Get a specific A/B test by ID.
   */
  async getABTest(testId: string): Promise<ABTest> {
    return this.client.request('GET', `/api/v1/evolution/ab-tests/${testId}`);
  }

  /** Get evolution overview. */
  async getOverview(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/evolution');
  }

  /** Get agent evolution dashboard timeline events. */
  async getAgentEvolutionTimeline(params?: AgentEvolutionTimelineOptions): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/agent-evolution/timeline', {
      params: params as Record<string, unknown>,
    });
  }

  /** Get agent evolution ELO trend data. */
  async getAgentEvolutionEloTrends(params?: AgentEvolutionEloTrendsOptions): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/agent-evolution/elo-trends', {
      params: params as Record<string, unknown>,
    });
  }

  /** Get pending agent evolution changes. */
  async getAgentEvolutionPending(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/agent-evolution/pending');
  }

  /** Approve a pending agent evolution change. */
  async approveAgentEvolutionChange(changeId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/agent-evolution/pending/${encodeURIComponent(changeId)}/approve`
    );
  }

  /** Reject a pending agent evolution change. */
  async rejectAgentEvolutionChange(changeId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'POST',
      `/api/v1/agent-evolution/pending/${encodeURIComponent(changeId)}/reject`
    );
  }
}

File: sdk/python/aragora_sdk/namespaces/evolution.py
"""
Evolution Namespace API

Provides access to agent evolution and A/B testing.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class EvolutionAPI:
    """Synchronous Evolution API for agent evolution tracking."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_history(self, agent: str) -> dict[str, Any]:
        """Get evolution history for an agent.

        Args:
            agent: Agent identifier.

        Returns:
            Evolution history with versions and metrics.
        """
        return self._client.request("GET", f"/api/v1/evolution/{agent}/history")

    def get_prompt(self, agent: str) -> dict[str, Any]:
        """Get current prompt for an agent.

        Args:
            agent: Agent identifier.

        Returns:
            Agent prompt configuration.
        """
        return self._client.request("GET", f"/api/v1/evolution/{agent}/prompt")

    def list_ab_tests(self) -> dict[str, Any]:
        """List active A/B tests.

        Returns:
            List of A/B tests with configurations.
        """
        return self._client.request("GET", "/api/v1/evolution/ab-tests")

    def get_patterns(self) -> dict[str, Any]:
        """Get evolution patterns.

        Returns:
            Detected evolution patterns.
        """
        return self._client.request("GET", "/api/v1/evolution/patterns")

    def get_summary(self) -> dict[str, Any]:
        """Get evolution summary.

        Returns:
            Summary of evolution metrics.
        """
        return self._client.request("GET", "/api/v1/evolution/summary")

    def get_overview(self) -> dict[str, Any]:
        """Get evolution overview.

        Returns:
            Dict with overall evolution status and metrics.
        """
        return self._client.request("GET", "/api/v1/evolution")

    def get_agent_evolution_timeline(
        self,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Get agent evolution dashboard timeline events."""
        return self._client.request(
            "GET",
            "/api/v1/agent-evolution/timeline",
            params={"limit": limit, "offset": offset},
        )

    def get_agent_evolution_elo_trends(self, period: str = "7d") -> dict[str, Any]:
        """Get agent evolution ELO trend data."""
        return self._client.request(
            "GET",
            "/api/v1/agent-evolution/elo-trends",
            params={"period": period},
        )

    def get_agent_evolution_pending(self) -> dict[str, Any]:
        """Get pending agent evolution changes."""
        return self._client.request("GET", "/api/v1/agent-evolution/pending")

    def approve_agent_evolution_change(self, change_id: str) -> dict[str, Any]:
        """Approve a pending agent evolution change."""
        return self._client.request(
            "POST",
            f"/api/v1/agent-evolution/pending/{change_id}/approve",
        )

    def reject_agent_evolution_change(self, change_id: str) -> dict[str, Any]:
        """Reject a pending agent evolution change."""
        return self._client.request(
            "POST",
            f"/api/v1/agent-evolution/pending/{change_id}/reject",
        )


class AsyncEvolutionAPI:
    """Asynchronous Evolution API for agent evolution tracking."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_history(self, agent: str) -> dict[str, Any]:
        """Get evolution history for an agent.

        Args:
            agent: Agent identifier.

        Returns:
            Evolution history with versions and metrics.
        """
        return await self._client.request("GET", f"/api/v1/evolution/{agent}/history")

    async def get_prompt(self, agent: str) -> dict[str, Any]:
        """Get current prompt for an agent.

        Args:
            agent: Agent identifier.

        Returns:
            Agent prompt configuration.
        """
        return await self._client.request("GET", f"/api/v1/evolution/{agent}/prompt")

    async def list_ab_tests(self) -> dict[str, Any]:
        """List active A/B tests.

        Returns:
            List of A/B tests with configurations.
        """
        return await self._client.request("GET", "/api/v1/evolution/ab-tests")

    async def get_patterns(self) -> dict[str, Any]:
        """Get evolution patterns.

        Returns:
            Detected evolution patterns.
        """
        return await self._client.request("GET", "/api/v1/evolution/patterns")

    async def get_summary(self) -> dict[str, Any]:
        """Get evolution summary.

        Returns:
            Summary of evolution metrics.
        """
        return await self._client.request("GET", "/api/v1/evolution/summary")

    async def get_overview(self) -> dict[str, Any]:
        """Get evolution overview."""
        return await self._client.request("GET", "/api/v1/evolution")

    async def get_agent_evolution_timeline(
        self,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Get agent evolution dashboard timeline events."""
        return await self._client.request(
            "GET",
            "/api/v1/agent-evolution/timeline",
            params={"limit": limit, "offset": offset},
        )

    async def get_agent_evolution_elo_trends(self, period: str = "7d") -> dict[str, Any]:
        """Get agent evolution ELO trend data."""
        return await self._client.request(
            "GET",
            "/api/v1/agent-evolution/elo-trends",
            params={"period": period},
        )

    async def get_agent_evolution_pending(self) -> dict[str, Any]:
        """Get pending agent evolution changes."""
        return await self._client.request("GET", "/api/v1/agent-evolution/pending")

    async def approve_agent_evolution_change(self, change_id: str) -> dict[str, Any]:
        """Approve a pending agent evolution change."""
        return await self._client.request(
            "POST",
            f"/api/v1/agent-evolution/pending/{change_id}/approve",
        )

    async def reject_agent_evolution_change(self, change_id: str) -> dict[str, Any]:
        """Reject a pending agent evolution change."""
        return await self._client.request(
            "POST",
            f"/api/v1/agent-evolution/pending/{change_id}/reject",
        )
